crates/wast/src/encode.rs

```diff
@@ -1,7 +1,19 @@
+use crate::core::Instruction;
+use crate::core::Func;
+use crate::core::BranchHintStruct;
+
 pub(crate) trait Encode {
     fn encode(&self, e: &mut Vec<u8>);
 }
 
+pub(crate) trait EncodeExpr {
+    fn encode_expr(&self, e: &mut Vec<u8>) -> Vec<(&Instruction<'_>, usize)>;
+}
+
+pub(crate) trait EncodeFunction {
+    fn encode_function(&self, e: &mut Vec<u8>) -> (&Func<'_>, Vec<BranchHintStruct>);
+}
```

Comment:
Also to avoid trait juggling and method juggling I might recommend throwing all of function encoding into "one big function" (or big-ish depending on how you'd like to slice it) given how intimately tied together the emission loop and keeping track of offsets is going to be